### PR TITLE
feat: Add dropped events/attributes/links counts to zipkin + jaeger exporters

### PR DESCRIPTION
**Pull request by ahayworth**
I was reviewing for spec compliance today, and noticed that we are not adding `otel.dropped_{events,attributes,links}_count` to span tags as required by the Zipkin and Jaeger exporter specs. We regret the omission.

https://github.com/open-telemetry/opentelemetry-specification/blob/f04897c16c9b9690076d52fdd94235e92ceefc96/specification/trace/sdk_exporters/jaeger.md
https://github.com/open-telemetry/opentelemetry-specification/blob/f04897c16c9b9690076d52fdd94235e92ceefc96/specification/trace/sdk_exporters/zipkin.md

### CODE REVIEW

**Comment by ahayworth** [comment]
Tests should pass after #1236 is merged.

**Review by arielvalentin — APPROVED** [review]

File: exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb [hunk 1]
```diff
@@ -64,10 +64,23 @@ def encoded_tag(key, value)
...
+          tags = encoded_tags(span_data.attributes) +
+                 encoded_status(span_data.status) +
+                 encoded_kind(span_data.kind) +
+                 encoded_instrumentation_library(span_data.instrumentation_library)
+
+          dropped_attributes_count = span_data.total_recorded_attributes - span_data.attributes&.size.to_i
+          dropped_events_count = span_data.total_recorded_events - span_data.events&.size.to_i
+          dropped_links_count = span_data.total_recorded_links - span_data.links&.size.to_i
+
+          tags << encoded_tag('otel.dropped_attributes_count', dropped_attributes_count) if dropped_attributes_count.positive?
```
[the remark this answers is not in the document]
Comment (ahayworth) [reply exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb:80]: Indeed. And I had actually tested it before with `if dropped_attributes_count > 0` and rubocop yelled at me until I switched to `positive?` (which works in this case too).

File: exporter/zipkin/test/opentelemetry/exporters/zipkin/transformer_test.rb [hunk 2]
```diff
@@ -59,6 +59,28 @@
     _(tags).must_equal('akey' => 'avalue', 'bar' => 'baz', 'otel.library.version' => '0.0.0', 'otel.library.name' => 'vendorlib')
   end
 
+  it 'records dropped attribute, event, and links counts when things were dropped' do
+    resource = OpenTelemetry::SDK::Resources::Resource.create('service.name' => 'foo')
+    span_data = create_span_data(total_recorded_attributes: 1, total_recorded_events: 1, total_recorded_links: 1)
+    encoded_span = Transformer.to_zipkin_span(span_data, resource)
+    tags = encoded_span['tags']
+    _(tags).must_equal(
+      'otel.library.version' => '0.0.0',
```
[the remark this answers is not in the document]
Comment (plantfansam) [reply exporter/zipkin/test/opentelemetry/exporters/zipkin/transformer_test.rb:68]: +1 to constantizing! I have no preference as to which drawer we should shove them in 😄

File: exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb [hunk 3]
```diff
@@ -64,10 +64,23 @@ def encoded_tag(key, value)
...
+        def encoded_span(span_data) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
           start_time = span_data.start_timestamp / 1_000
           duration = span_data.end_timestamp / 1_000 - start_time
 
+          tags = encoded_tags(span_data.attributes) +
+                 encoded_status(span_data.status) +
+                 encoded_kind(span_data.kind) +
+                 encoded_instrumentation_library(span_data.instrumentation_library)
+
+          dropped_attributes_count = span_data.total_recorded_attributes - span_data.attributes&.size.to_i
```
[the remark this answers is not in the document]
Comment (plantfansam) [reply exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb:76]: To clarify, I didn't mean to say that these computations had already been done, but rather that `.total_recorded_attributes` etc. were fields on `SpanData`. No change needed!

File: exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb [hunk 4]
```diff
@@ -64,10 +64,23 @@ def encoded_tag(key, value)
           )
         end
 
-        def encoded_span(span_data) # rubocop:disable Metrics/AbcSize
+        def encoded_span(span_data) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
           start_time = span_data.start_timestamp / 1_000
           duration = span_data.end_timestamp / 1_000 - start_time
 
+          tags = encoded_tags(span_data.attributes) +
```
[the remark this answers is not in the document]
Comment (ahayworth) [reply exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb:71]: Indeed, it just felt really unwieldy not to separate it. Of course, that made the 👮‍♂️ start complaining... 🙄

File: exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahayworth) [reply exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb:76]: Ah gotcha, yes that makes sense. Whoo, communication! 😆

File: exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb [hunk 5]
```diff
@@ -64,10 +64,23 @@ def encoded_tag(key, value)
...
+                 encoded_status(span_data.status) +
+                 encoded_kind(span_data.kind) +
+                 encoded_instrumentation_library(span_data.instrumentation_library)
+
+          dropped_attributes_count = span_data.total_recorded_attributes - span_data.attributes&.size.to_i
+          dropped_events_count = span_data.total_recorded_events - span_data.events&.size.to_i
+          dropped_links_count = span_data.total_recorded_links - span_data.links&.size.to_i
```
[the remark this answers is not in the document]
Comment (ahayworth) [reply exporter/jaeger/lib/opentelemetry/exporter/jaeger/encoder.rb:78]: Yes, I'm virtually certain. After looking through the places where we modify the variables you pointed out, we're only ever mathematically modifying them. Anything that causes it to be `nil` in the future would be a bug we introduced - and should hopefully implode the test suite and never be released.

File: exporter/zipkin/test/opentelemetry/exporters/zipkin/transformer_test.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahayworth) [reply exporter/zipkin/test/opentelemetry/exporters/zipkin/transformer_test.rb:68]: Aaaaaactually I would like to punt; we do this in a few other places. I opened #1250 to track it, and I think we could shove them into `opentelemetry-semantic_conventions` later on down the line. Do we feel okay about that?

[the remark this answers is not in the document]
Comment (plantfansam) [reply exporter/zipkin/test/opentelemetry/exporters/zipkin/transformer_test.rb:68]: I feel very OK about punting and addressing in #1250

**Review by fbogsany — APPROVED** [review]

**Pull request merged**